mantidproject/mantid
# 15703 inelastic form factor correction

This PR adds a new algorithm which corrects a workspace with a |Q| axis for the magnetic form factor. That is, given an ion name, it obtains the form factor F(Q) from `MagneticIon` and scales the input workspace by 1/|F(Q)|^2. 

**To test:**

Using the attached files (data of PrAl3 measured on MARI):
[magffcorr_files.zip](https://github.com/mantidproject/mantid/files/470176/magffcorr_files.zip)

You can run this in MantidPlot: 

```
from Direct import DirectEnergyConversion
from mantid.simpleapi import *
rd = DirectEnergyConversion.DirectEnergyConversion('MARI')
ws = rd.convert_to_energy(21334, 21337, 15, [-15,0.05,15], 'mari_res2013.map',
    monovan_run=21346, sample_mass=10.62, sample_rmm=221.856, monovan_mapfile='mari_res2013.map')
ws_sqw = SofQW3(ws, [0,0.05,6], 'Direct', 15)
Uncorrected = Rebin2D(ws_sqw, '4,1,5', '0.7,0.05,4.3', True, True)
Uncorrected = Uncorrected / np.max(Uncorrected.readY(0))
ws_corr = MagFormFactorCorrection(ws_sqw, IonName='Pr3', FormFactorWorkspace='Pr3FF')
Corrected = Rebin2D(ws_corr[0], '3.5,3,5.5', '0.7,0.05,4.3', True, True)
Corrected = Corrected / np.max(Corrected.readY(0))
Pr3_FofQsquared = mtd['Pr3FF'] * mtd['Pr3FF']
plotSpectrum([Uncorrected, Corrected, Pr3_FofQsquared], 0, error_bars=True)
```

You should get this picture:
![magffcorr](https://cloud.githubusercontent.com/assets/14106963/18569075/5b64eb44-7b98-11e6-8293-fbcc1a9940c1.png)

Or, using a fake dataset:

```
import numpy as np
ws = CreateSampleWorkspace(binWidth = 0.1, XMin = 0, XMax = 50, XUnit = 'DeltaE')
ws = ScaleX(ws, -15, "Add")
LoadInstrument(ws, InstrumentName='MARI', RewriteSpectraMap = True)
ws = SofQW(ws, [0, 0.05, 8], 'Direct', 35)
Q = ws.getAxis(1).extractValues()
for i in range(len(Q)-1):
    qv = ( (Q[i]+Q[i+1])*0.5 ) / 4 / np.pi
    y = ws.dataY(i)
    y *= np.exp(-16*qv*qv)
ws_corr = MagFormFactorCorrection(ws, IonName='Fe3', FormFactorWorkspace='Fe3FF')
wsl = Rebin2D(ws, '8.5,3,11.5', '1.2,0.05,7.3', True, True)
wsl = wsl / np.max(wsl.readY(0))
wsf = Rebin2D(ws_corr[0], '8.5,3,11.5', '1.2,0.05,7.3', True, True)
wsf = wsf / np.max(wsf.readY(0))
Fe3FF2 = mtd['Fe3FF'] * mtd['Fe3FF']
plotSpectrum([wsl, wsf, Fe3FF2],0,error_bars=True)
```

Check that the output corrected workspace (`ws_corr` of the cut `wsf`) has nearly uniform intensity 
(this doesn't work very well for the real data example, PrAl3, possibly because currently we're using only the spin-only formula for the form factor, whereas the rare earths need to also include the second order integrals - fixed by #17458 - so maybe it will be better in future. It could also be the data being problematic... Update: now works well, after #17458 was merged)



Fixes #15703 .



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
